File: frontend/src/api/avicultureHealth.ts
/**
 * Aviculture Health API (Module 15, Part 6).
 *
 * Individual-bird health records, quarantine, disease events and analytics.
 * Analytics arrive honesty-labelled from the pure health engine. Diagnostic
 * records carry a veterinary-guidance disclaimer — ARIA explains, never diagnoses.
 *
 * Distinct from the platform's flock ``health.ts`` (Module 1) — different domain.
 */
import apiClient from "./client";
import type { Labelled } from "./aviaries";

type APISuccess<T> = { data: T };

const base = (farmId: string) => `/farms/${farmId}/aviculture`;

export interface HealthRecord {
  id: string; bird_id: string; record_type: string; recorded_on: string; title: string | null;
  summary: string | null; weight_grams: string | null; body_condition: string | null;
  status: string; severity: string | null; next_due_on: string | null; details: Record<string, unknown>;
}

export interface Quarantine {
  id: string; farm_id: string; bird_id: string; aviary_id: string | null; reason: string | null;
  started_on: string; expected_end_on: string | null; ended_on: string | null; status: string; outcome: string | null;
}

export interface DiseaseEvent {
  id: string; farm_id: string; disease_name: string; status: string; started_on: string;
  resolved_on: string | null; affected_count: number; is_notifiable: boolean; notes: string | null;
}

export interface WeightTrend {
  bird_id: string;
  trend: {
    count: Labelled<number>; latest: Labelled<number | null>; first?: Labelled<number | null>;
    change_grams: Labelled<number | null>; direction: Labelled<string | null>;
    series?: { date: string; grams: number }[];
  };
}

export interface HealthSummary {
  records_total: Labelled; records_by_type: Record<string, number>; active_birds: Labelled;
  mortality_rate_pct: Labelled; vaccination_coverage_pct: Labelled;
  active_quarantines: Labelled; active_disease_events: Labelled;
  preventive_due: { due_count: Labelled; overdue_count: Labelled };
}

export async function listHealthRecords(farmId: string, birdId: string): Promise<HealthRecord[]> {
  const { data } = await apiClient.get<APISuccess<HealthRecord[]>>(`${base(farmId)}/birds/${birdId}/health`);
  return data.data;
}
export async function addHealthRecord(farmId: string, birdId: string, body: {
  record_type: string; recorded_on?: string; title?: string; summary?: string; weight_grams?: string;
  severity?: string; next_due_on?: string; details?: Record<string, unknown>;
}): Promise<HealthRecord> {
  const { data } = await apiClient.post<APISuccess<HealthRecord>>(`${base(farmId)}/birds/${birdId}/health`, body);
  return data.data;
}
export async function getWeightTrend(farmId: string, birdId: string): Promise<WeightTrend> {
  const { data } = await apiClient.get<APISuccess<WeightTrend>>(`${base(farmId)}/birds/${birdId}/weight-trend`);
  return data.data;
}
export async function startQuarantine(farmId: string, birdId: string, body: { reason?: string; started_on?: string; expected_end_on?: string }): Promise<Quarantine> {
  const { data } = await apiClient.post<APISuccess<Quarantine>>(`${base(farmId)}/birds/${birdId}/quarantine`, body);
  return data.data;
}
export async function listBirdQuarantine(farmId: string, birdId: string): Promise<Quarantine[]> {
  const { data } = await apiClient.get<APISuccess<Quarantine[]>>(`${base(farmId)}/birds/${birdId}/quarantine`);
  return data.data;
}
export async function listQuarantine(farmId: string, activeOnly = false): Promise<Quarantine[]> {
  const { data } = await apiClient.get<APISuccess<Quarantine[]>>(`${base(farmId)}/quarantine`, { params: { active_only: activeOnly } });
  return data.data;
}
export async function releaseQuarantine(farmId: string, quarantineId: string, outcome?: string): Promise<Quarantine> {
  const { data } = await apiClient.post<APISuccess<Quarantine>>(`${base(farmId)}/quarantine/${quarantineId}/release`, { outcome });
  return data.data;
}
export async function listDiseaseEvents(farmId: string): Promise<DiseaseEvent[]> {
  const { data } = await apiClient.get<APISuccess<DiseaseEvent[]>>(`${base(farmId)}/disease-events`);
  return data.data;
}
export async function createDiseaseEvent(farmId: string, body: { disease_name: string; status?: string; affected_count?: number; is_notifiable?: boolean }): Promise<DiseaseEvent> {
  const { data } = await apiClient.post<APISuccess<DiseaseEvent>>(`${base(farmId)}/disease-events`, body);
  return data.data;
}
export async function updateDiseaseEvent(farmId: string, eventId: string, body: { status?: string; affected_count?: number }): Promise<DiseaseEvent> {
  const { data } = await apiClient.patch<APISuccess<DiseaseEvent>>(`${base(farmId)}/disease-events/${eventId}`, body);
  return data.data;
}
export async function getHealthSummary(farmId: string): Promise<HealthSummary> {
  const { data } = await apiClient.get<APISuccess<HealthSummary>>(`${base(farmId)}/health/summary`);
  return data.data;
}

File: backend/app/schemas/aviculture_health.py
"""
Greena — Aviculture Health Schemas (Module 15, Part 6)

Contracts for individual-bird health records, quarantine, disease events and
analytics. Differential diagnoses are carried inside ``details`` and always
returned with a disclaimer stamped by the service — veterinary guidance, never a
definitive diagnosis (Doc 04 §7, Doc 06/Part 6 spec).
"""

from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import Field, field_validator

from app.models import aviculture as avi
from app.schemas.aviculture import _one_of
from app.schemas.base import AGRIOSSchema, TimestampedSchema


# ── Health records ────────────────────────────────────────────────────────────

class HealthRecordCreate(AGRIOSSchema):
    record_type: str = Field("observation")
    recorded_on: date | None = None
    title: str | None = Field(None, max_length=255)
    summary: str | None = None
    weight_grams: Decimal | None = Field(None, ge=0)
    body_condition: str | None = Field(None, max_length=30)
    status: str = Field("recorded")
    severity: str | None = None
    next_due_on: date | None = None
    details: dict = Field(default_factory=dict)

    _t = field_validator("record_type")(_one_of("record_type", avi.HEALTH_RECORD_TYPE_VALUES))
    _s = field_validator("status")(_one_of("status", avi.HEALTH_RECORD_STATUS_VALUES))
    _sev = field_validator("severity")(_one_of("severity", avi.HEALTH_SEVERITY_VALUES))


class HealthRecordUpdate(AGRIOSSchema):
    status: str | None = None
    severity: str | None = None
    next_due_on: date | None = None
    summary: str | None = None

    _s = field_validator("status")(_one_of("status", avi.HEALTH_RECORD_STATUS_VALUES))
    _sev = field_validator("severity")(_one_of("severity", avi.HEALTH_SEVERITY_VALUES))


class HealthRecordResponse(TimestampedSchema):
    bird_id: UUID
    record_type: str
    recorded_on: date
    title: str | None
    summary: str | None
    weight_grams: Decimal | None
    body_condition: str | None
    status: str
    severity: str | None
    next_due_on: date | None
    details: dict


# ── Quarantine ────────────────────────────────────────────────────────────────

class QuarantineStart(AGRIOSSchema):
    reason: str | None = Field(None, max_length=255)
    aviary_id: UUID | None = None
    started_on: date | None = None
    expected_end_on: date | None = None
    notes: str | None = None


class QuarantineEnd(AGRIOSSchema):
    ended_on: date | None = None
    outcome: str | None = Field(None, max_length=255)


class QuarantineResponse(TimestampedSchema):
    farm_id: UUID
    bird_id: UUID
    aviary_id: UUID | None
    reason: str | None
    started_on: date
    expected_end_on: date | None
    ended_on: date | None
    status: str
    outcome: str | None
    notes: str | None


# ── Disease events ────────────────────────────────────────────────────────────

class DiseaseEventCreate(AGRIOSSchema):
    disease_name: str = Field(..., min_length=1, max_length=200)
    aviary_id: UUID | None = None
    status: str = Field("suspected")
    started_on: date | None = None
    affected_count: int = Field(0, ge=0)
    is_notifiable: bool = False
    notes: str | None = None

    _s = field_validator("status")(_one_of("status", avi.DISEASE_EVENT_STATUS_VALUES))


class DiseaseEventUpdate(AGRIOSSchema):
    status: str | None = None
    resolved_on: date | None = None
    affected_count: int | None = Field(None, ge=0)
    notes: str | None = None

    _s = field_validator("status")(_one_of("status", avi.DISEASE_EVENT_STATUS_VALUES))


class DiseaseEventResponse(TimestampedSchema):
    farm_id: UUID
    aviary_id: UUID | None
    disease_name: str
    status: str
    started_on: date
    resolved_on: date | None
    affected_count: int
    is_notifiable: bool
    notes: str | None


# ── Analytics ─────────────────────────────────────────────────────────────────

class WeightTrendResponse(AGRIOSSchema):
    bird_id: UUID
    trend: dict


class HealthSummaryResponse(AGRIOSSchema):
    records_total: dict
    records_by_type: dict
    active_birds: dict
    mortality_rate_pct: dict
    vaccination_coverage_pct: dict
    active_quarantines: dict
    active_disease_events: dict
    preventive_due: dict
